File: algorithms/lua/sorting/quick_sort.lua
function quicksort(arr)
    if #arr <= 1 then
        return arr
    end

    local pivot = arr[#arr]
    table.remove(arr, #arr)

    local left = {}
    local right = {}

    for _, v in ipairs(arr) do
        if v <= pivot then
            table.insert(left, v)
        else
            table.insert(right, v)
        end
    end

    local sorted_left = quicksort(left)
    local sorted_right = quicksort(right)

    local result = {}
    for _, v in ipairs(sorted_left) do table.insert(result, v) end
    table.insert(result, pivot)
    for _, v in ipairs(sorted_right) do table.insert(result, v) end

    return result
end

-- Example usage
local arr = {10, 7, 8, 9, 1, 5}
print("Original array: " .. table.concat(arr, ", "))
local sorted_arr = quicksort(arr)
print("Sorted array: " .. table.concat(sorted_arr, ", "))

File: algorithms/scala/sorting/QuickSort.scala
object QuickSort {
  def quickSort(arr: Array[Int]): Array[Int] = {
    if (arr.length <= 1) arr
    else {
      val pivot = arr(arr.length / 2)
      Array.concat(
        quickSort(arr filter (pivot >)),
        arr filter (pivot ==),
        quickSort(arr filter (pivot <))
      )
    }
  }

  def main(args: Array[String]): Unit = {
    val arr = Array(10, 7, 8, 9, 1, 5)
    println("Original array: " + arr.mkString(", "))
    val sortedArr = quickSort(arr)
    println("Sorted array: " + sortedArr.mkString(", "))
  }
}
